libyara/modules/macho.c

```diff
@@ -0,0 +1,1180 @@
+    set_integer(yr_##bo##32toh(header_64->reserved), object, "reserved");      \
+  }                                                                            \
+                                                                               \
+  uint64_t num_sg = 0;                                                         \
+  uint8_t *command = (uint8_t*)(header + 1);                                   \
+  for (unsigned i = 0; i < yr_##bo##32toh(header->ncmds); ++i)                 \
+  {                                                                            \
+    load_command_t* command_struct = (load_command_t*)command;                 \
+    switch(yr_##bo##32toh(command_struct->cmd))                                \
+    {                                                                          \
+      case LC_SEGMENT:                                                         \
+      case LC_SEGMENT_64:                                                      \
+      {                                                                        \
+        macho_handle_segments_sections_##bits##_##bo(command, num_sg, object); \
+        num_sg++;                                                              \
+        break;                                                                 \
+      }                                                                        \
+      case LC_UNIXTHREAD:                                                      \
+      {                                                                        \
+        macho_handle_unixthread_##bo(command, object, context);                \
+        break;                                                                 \
+      }                                                                        \
+      case LC_MAIN:                                                            \
+      {                                                                        \
+        macho_handle_main_##bo(command, object, context);                      \
+        break;                                                                 \
+      }                                                                        \
+    }                                                                          \
+    command += yr_##bo##32toh(command_struct->cmdsize);                        \
+  }                                                                            \
+                                                                               \
+  set_integer(num_sg, object, "number_of_segments");                           \
+}                                                                              \
+
+MACHO_PARSE_FILE(32,le)
+MACHO_PARSE_FILE(64,le)
+MACHO_PARSE_FILE(32,be)
+MACHO_PARSE_FILE(64,be)
+
+
+// Parse Mach-O file.
+
+void macho_parse_file(
+    void* data,
+    YR_OBJECT* object,
+    YR_SCAN_CONTEXT* context)
+{
+  uint8_t* magic = (uint8_t*)data;
```

Comment:
i cant see any boundary checks, without trying this pr, it feels like this code will crash when parsing truncated files